kaldb/src/main/java/com/slack/kaldb/logstore/opensearch/OpenSearchAggregationAdapter.java

```diff
@@ -0,0 +1,351 @@
+            singletonMap(
+                "default",
+                new NamedAnalyzer("default", AnalyzerScope.INDEX, new StandardAnalyzer())),
+            emptyMap(),
+            emptyMap()),
+        new NamedXContentRegistry(ClusterModule.getNamedXWriteables()),
+        similarityService,
+        new IndicesModule(emptyList()).getMapperRegistry(),
+        () -> {
+          throw new UnsupportedOperationException();
+        },
+        () -> true,
+        null);
+  }
+
+  /**
+   * Minimal implementation of an OpenSearch QueryShardContext while still allowing an
+   * AggregatorFactory to successfully instantiate. See AggregatorFactory.class
+   */
+  private static QueryShardContext buildQueryShardContext(
+      BigArrays bigArrays,
+      IndexSettings indexSettings,
+      SimilarityService similarityService,
+      MapperService mapperService) {
+    final ValuesSourceRegistry valuesSourceRegistry = buildValueSourceRegistry();
+    return new QueryShardContext(
+        0,
+        indexSettings,
+        bigArrays,
+        null,
+        new IndexFieldDataService(
+                indexSettings,
+                new IndicesFieldDataCache(
+                    indexSettings.getSettings(), new IndexFieldDataCache.Listener() {}),
+                new NoneCircuitBreakerService(),
+                mapperService)
+            ::getForField,
+        mapperService,
+        similarityService,
+        null,
+        null,
+        null,
+        null,
+        null,
+        () -> Instant.now().toEpochMilli(),
+        null,
+        s -> false,
+        null,
+        valuesSourceRegistry);
+  }
+
+  /**
+   * Registers a field type and name to the MapperService for use in aggregations. This informs the
+   * aggregators how to access a specific field and what value type it contains. registerField(
+   * mapperService, LogMessage.SystemField.TIME_SINCE_EPOCH.fieldName, b -> b.field("type",
+   * "long"));
+   */
+  private static void registerField(
+      MapperService mapperService,
+      String fieldName,
+      CheckedConsumer<XContentBuilder, IOException> buildField)
+      throws IOException {
+    XContentBuilder mapping = fieldMapping(fieldName, buildField);
+    mapperService.merge(
+        "_doc",
+        new CompressedXContent(BytesReference.bytes(mapping)),
+        MapperService.MergeReason.MAPPING_UPDATE);
+  }
+
+  /**
+   * Builds a Lucene collector that can be used in native Lucene search methods using the OpenSearch
+   * aggregations implementation.
+   */
+  public static Aggregator buildAutoDateHistogramAggregator(int numBuckets) throws IOException {
+    IndexSettings indexSettings = buildIndexSettings();
+    SimilarityService similarityService = new SimilarityService(indexSettings, null, emptyMap());
+    MapperService mapperService = buildMapperService(indexSettings, similarityService);
+
+    registerField(
+        mapperService,
+        LogMessage.SystemField.TIME_SINCE_EPOCH.fieldName,
```

Comment:
How does "long" translate to from our schema?

Reply:
This would map to the same ones you can put into OpenSearch using the mapping API, which the entirety is listed here - https://opensearch.org/docs/2.1/opensearch/supported-field-types/index/. For example, a date format would look something like the following:

```
b -> b.field("type", "date")
      .field("format", "E, d MMM yyyy HH:mm:ss Z")
      .field("locale", "de")
```

In the case of date, it's actually indexed as a long (https://opensearch.org/docs/2.1/opensearch/supported-field-types/date/), and then the mapping informs how to resolve this into a string when outputting the final format.